src/coprocessor/codec/convert.rs

```diff
@@ -43,53 +101,163 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
 }
 
-/// `convert_float_to_int` converts an f64 value to an i64 value.
-///  Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_int(
-    fval: f64,
-    lower_bound: i64,
-    upper_bound: i64,
-    tp: FieldTypeTp,
-) -> Result<i64> {
-    // TODO any performance problem to use round directly?
+/// Converts an f64 value to an i64 value.
+#[inline]
+pub fn convert_float_to_int(ctx: &mut EvalContext, fval: f64, tp: FieldTypeTp) -> Result<i64> {
     let val = fval.round();
+    let lower_bound = integer_signed_lower_bound(tp);
     if val < lower_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
     }
 
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as f64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
-/// `convert_float_to_uint` converts a f64 value to a u64 value.
-/// Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_uint(fval: f64, upper_bound: u64, tp: FieldTypeTp) -> Result<u64> {
-    // TODO any performance problem to use round directly?
+/// Converts a byte arrays to an i64 in best effort.
+pub fn convert_bytes_to_int(ctx: &mut EvalContext, bytes: &[u8], tp: FieldTypeTp) -> Result<i64> {
+    let s = match str::from_utf8(bytes) {
+        Ok(s) => s,
+        Err(err) => {
+            ctx.handle_truncate(true)?;
+            let (valid, _) = bytes.split_at(err.valid_up_to());
+            unsafe { str::from_utf8_unchecked(valid) }
+        }
+    };
+    let s = s.trim();
+    let vs = get_valid_int_prefix(ctx, s)?;
+    let val = vs.parse::<i64>();
+    match val {
+        Ok(val) => convert_int_to_int(ctx, val, tp),
+        Err(_) => {
+            ctx.handle_overflow(Error::overflow("BIGINT", &vs))?;
+            let val = if vs.starts_with('-') {
+                integer_signed_lower_bound(tp)
+            } else {
+                integer_signed_upper_bound(tp)
+            };
+            Ok(val)
+        }
+    }
+}
+
+/// Converts a `DateTime` to an i64 value
+#[inline]
+pub fn convert_datetime_to_int(
+    ctx: &mut EvalContext,
+    dt: &DateTime,
+    tp: FieldTypeTp,
+) -> Result<i64> {
+    // TODO: avoid this clone after refactor the `Time`
+    let mut t = dt.clone();
+    t.round_frac(DEFAULT_FSP)?;
+    let val = t.to_decimal()?.as_i64_with_ctx(ctx)?;
+    convert_int_to_int(ctx, val, tp)
+}
+
+/// Converts a `Duration` to an i64 value
+#[inline]
+pub fn convert_duration_to_int(
+    ctx: &mut EvalContext,
+    dur: Duration,
+    tp: FieldTypeTp,
+) -> Result<i64> {
+    // It's OK to clone because duration only occupies 8 bytes after #4858 merged
+    let dur = dur.round_frac(DEFAULT_FSP)?;
+    let val = Decimal::try_from(dur)?.as_i64_with_ctx(ctx)?;
+    convert_int_to_int(ctx, val, tp)
```

Comment:
Same issue as above. Accepting ctx multiple times may result in multiple warnings being generated in a single UDF function call?